### PR TITLE
Add support for VMs with multiple IB devices

### PR DESCRIPTION
**Pull request by longlimsft**
Newer HPC VM supports multiple Infiniband NICs. A new KVP format is
introduced in KVP pool0 to pass multiple IPs from KVP daemon. The KVP
is of format: (e.g passing two IPs to two IB)

key name: IPoIB_Data
value: NUMPAIRS:2|FFFFFFFFFFF1:172.16.0.1|FFFFFFFFFFF2:172.16.0.2

FFFFFFFFFFF1 and FFFFFFFFFFF2 are the MAC address for the IP over IB device.
172.16.0.1 and 172.16.0.2 are their respective IPs.

If this key is not present in the KVP pool0, the Linux agent falls back to
the old code path of provisioning a single IB device, thus maintaining
compatiblity with older HPC VMs.

## Description

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by narrieta** [comment]
running automation

**Comment by narrieta** [comment]
automation ok

**Review by narrieta — DISMISSED** [review]

File: azurelinuxagent/common/rdma.py [hunk 1]
```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
...
+        kvp_value_size = 2048
+        driver_info_source = '/var/lib/hyperv/.kvp_pool_0'
+
+        if not os.path.isfile(driver_info_source):
+            logger.error("RDMA: can't read KVP pool 0")
+            return (None, None)
+
+        key_0 = None
+        value_0 = None
+        f = open(driver_info_source, "rb")  # pylint: disable=C0103
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:297]: Please take care of the pylint warning (#C0103) [rename `f` to a more meaningful name like `driver_file` or something)]

File: azurelinuxagent/common/rdma.py [hunk 2]
```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
...
+
+        net_dir = "/sys/class/net"
+        nics = os.listdir(net_dir)
+        count = 0
+
+        for nic in nics:
+            # look for IBoIP interface of format ibXXX
+            if not re.match(r"ib\d+", nic):
+                continue
+            address_file = open(net_dir + "/" + nic + "/address")
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:373]: NIT: consider using `os.path.join(net_dir, nic, "address")`

File: azurelinuxagent/common/rdma.py [hunk 3]
```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
...
+            mac_addr = address_file.read()
+            address_file.close()
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:375]: Wrap this in a `finally` clause to ensure the file is always closed 
(Consider using `with open(filename)` instead as that ensures that the file is always closed even on error - Eg: https://cmdlinetips.com/2016/01/opening-a-file-in-python-using-with-statement/)

File: azurelinuxagent/common/rdma.py [hunk 4]
```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
...
+                logger.error("RDMA: failed to parse address for device {0} address {1}".format(nic, mac_addr))
+                continue
+
+            # format an MAC address without :
+            mac_addr = ""
+            mac_addr = mac_addr.join(match.groups(0))
+
+            for mac_ip in mac_ip_array:
+                if mac_ip[0] == mac_addr:
+                    ret = shellutil.run("ip addr add {0}/16 dev {1}".format(mac_ip[1], nic))
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:394]: We're deprecating the use of `shellutil.run()` function. Please use `shellutil.run_command()` here instead

File: azurelinuxagent/common/rdma.py [hunk 5]
```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
...
+                if key_0 :
+                    key_0 = key_0.decode()
+                if key_0 == "IPoIB_Data":
+                    value_0 = value.partition(b"\x00")[0]
+                    if value_0 :
+                        value_0 = value_0.decode()
+                    break
+            else:
+                break
+        f.close()
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:312]: Please wrap this code in `try:except:finally` to ensure the file is always closed even on errors

**Review by larohra — CHANGES_REQUESTED** [review]
Added few comments

**Comment by larohra** [comment]
Hey @longlimsft, any update on the PR? Feel free to ping me if you need any clarification about any of my comments above!

File: azurelinuxagent/common/rdma.py (hunk 4 above)
Comment (longlimsft) [reply azurelinuxagent/common/rdma.py:394]: shellutil.run() will not raise an exception on error. Is there a way to not to raise an exception with shellutil.run_command() and return error code?

Comment (larohra) [reply azurelinuxagent/common/rdma.py:394]: you can just catch it and capture the return code like that. Check out the implementation here for an example - https://github.com/Azure/WALinuxAgent/blob/5e97cf23b53d53c95247e06ff3c93a85029271a3/azurelinuxagent/common/osutil/redhat.py#L111

Basically something like - 
```python
        ret_code = 0
        try:
            return shellutil.run_command(cmd, log_error=True)
        except shellutil.CommandError as error: 
            ret_code = error.returncode


```

**Comment by longlimsft** [comment]
> Hey @longlimsft, any update on the PR? Feel free to ping me if you need any clarification about any of my comments above!

I have updated the pull request. Please take a look.

File: azurelinuxagent/common/rdma.py [hunk 6]
```diff
@@ -265,11 +265,151 @@ def provision_network_direct_rdma(self):
...
+                    try:
+                        ip_command = ["ip", "addr", "add", "{0}/16".format(mac_ip[1]), "dev", nic]
+                        shellutil.run_command(ip_command)
+                    except shellutil.CommandError as error:
+                        ret = error.returncode
+
+                    if ret == 0:
+                        logger.info("RDMA: set address {0} to device {1}".format(mac_ip[1], nic))
+
+                    if ret and ret != 2:
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:404]: NIT: Redundant check - `if ret`, `ret` will always be defined in this scope (line 394 and 399)

**Review by larohra — COMMENTED** [review]
1 very small comment else LGTM. Will run DCR automation

**Review by larohra — APPROVED** [review]
Automation ok, LGTM

**Review by narrieta — APPROVED** [review]

**Pull request merged**